ledger/src/blockstore.rs

```diff
@@ -656,20 +656,22 @@ impl Blockstore {
         slot: Slot,
         erasure_meta: &ErasureMeta,
         available_shreds: &mut Vec<Shred>,
-        prev_inserted_codes: &mut HashMap<(u64, u64), Shred>,
+        prev_inserted_codes: &HashMap<(u64, u64), Shred>,
         code_cf: &LedgerColumn<cf::ShredCode>,
     ) {
         (erasure_meta.set_index..erasure_meta.set_index + erasure_meta.config.num_coding() as u64)
             .for_each(|i| {
                 if let Some(shred) = prev_inserted_codes
-                    .remove(&(slot, i))
+                    // Keeping the erasure coding shred in the previously inserted code
+                    // to allow us serving coding shreds over repair.
+                    .get(&(slot, i))
                     .map(|s| {
                         // Remove from the index so it doesn't get committed. We know
                         // this is safe to do because everything in
                         // `prev_inserted_codes` does not yet exist in blockstore
                         // (guaranteed by `check_cache_coding_shred`)
                         index.coding_mut().set_present(i, false);
```

Comment:
Looking at this again, and comparing with the how data-shreds are looked up here:
https://github.com/solana-labs/solana/blob/140a5f633/ledger/src/blockstore.rs#L636-L637
looks like data-shreds are also removed from `prev_inserted_datas` (not quite sure why).
But just based on the comparison of the two function, it seems like the difference is the
```rust
index.coding_mut().set_present(i, false);
```
line, and that is the part we need to remove.

Reply:
Thanks for pointing this out.  I actually discussed this with @carllin previously, and he suggests keeping `set_present(i, false)` but deleting `remove` instead.  Let me try to explain this (@carllin: please feel free to correct me if I say anything wrong :p).

So first, just_inserted_coding_shreds (i.e., prev_inserted_coding) is loaded here:
https://github.com/yhchiang-sol/solana/blob/data-shred/ledger/src/blockstore.rs#L914
which will also make the loaded coding shreds present:
https://github.com/yhchiang-sol/solana/blob/data-shred/ledger/src/blockstore.rs#L1206-L1211

And this diff keeps it untouched by making it immutable during the process:
https://github.com/yhchiang-sol/solana/blob/data-shred/ledger/src/blockstore.rs#L935

As we don't delete any coding shreds during the recovery, all the coding shreds will be inserted, which will then later be read in the following line next time during the recovery:
https://github.com/yhchiang-sol/solana/blob/data-shred/ledger/src/blockstore.rs#L914